lib/internal/modules/package_json_reader.js

```diff
@@ -1,30 +1,101 @@
 'use strict';
 
-const { SafeMap } = primordials;
+const {
+  JSONParse,
+  ObjectPrototypeHasOwnProperty,
+  SafeMap,
+} = primordials;
+const {
+  ERR_INVALID_PACKAGE_CONFIG,
+} = require('internal/errors').codes;
 const { internalModuleReadJSON } = internalBinding('fs');
-const { pathToFileURL } = require('url');
 const { toNamespacedPath } = require('path');
+const { kEmptyObject } = require('internal/util');
+const { fileURLToPath, pathToFileURL } = require('internal/url');
 
 const cache = new SafeMap();
 
 let manifest;
 
 /**
- *
  * @param {string} jsonPath
+ * @param {{
+ *   base?: string,
+ *   specifier: string,
+ *   isESM: boolean,
+ * }} options
+ * @returns {{
+ *   exists: boolean,
+ *   pjsonPath: string,
+ *   name?: string,
+ *   main?: string,
+ *   exports?: string | Record<string, unknown>,
+ *   imports?: string | Record<string, unknown>,
+ *   type: 'commonjs' | 'module' | 'none',
+ * }}
  */
-function read(jsonPath) {
+function read(jsonPath, { base, specifier, isESM } = kEmptyObject) {
   if (cache.has(jsonPath)) {
     return cache.get(jsonPath);
   }
 
-  const { 0: string, 1: containsKeys } = internalModuleReadJSON(
+  const string = internalModuleReadJSON(
     toNamespacedPath(jsonPath),
   );
-  const result = { string, containsKeys };
-  const { getOptionValue } = require('internal/options');
+  const result = {
+    __proto__: null,
+    exists: false,
+    pjsonPath: jsonPath,
+    main: undefined,
+    name: undefined,
+    type: 'none',
+    exports: undefined,
+    imports: undefined,
+  };
+
   if (string !== undefined) {
+    let parsed;
+    try {
+      parsed = JSONParse(string);
+    } catch (error) {
+      if (isESM) {
+        throw new ERR_INVALID_PACKAGE_CONFIG(
+          jsonPath,
+          (base ? `"${specifier}" from ` : '') + fileURLToPath(base || specifier),
+          error.message,
+        );
+      } else {
+        error.path = jsonPath;
+        error.message = 'Error parsing ' + jsonPath + ': ' + error.message;
+        throw error;
+      }
+    }
+
+    result.exists = true;
+
+    // ObjectPrototypeHasOwnProperty is used to avoid prototype pollution.
+    if (ObjectPrototypeHasOwnProperty(parsed, 'name') && typeof parsed.name === 'string') {
+      result.name = parsed.name;
+    }
```

Comment:
Nit: I think you can avoid the need for all these involved "own" checks by first setting `ObjectSetPrototypeOf(parsed.prototype, null)`

Reply:
altho i think that might indeed cause a deopt, it’d be much safer.

Reply:
Actually, `JSONParse` is a primordial—is it already pollution-safe?

Reply:
It produces a normal object, so no.

Reply:
I believe this causes deopt. Since, it is a `nit` I'm resolving this, if it's ok with everyone?

Reply:
It's okay with me as the author of the nit 🙂

PS i'm curious if it is a net deop compared with the cost of ~5 own checks 🧐